src/test/java/io/vertx/ext/consul/instance/ConsulInstance.java

```diff
@@ -0,0 +1,304 @@
+package io.vertx.ext.consul.instance;
+
+import com.github.dockerjava.api.model.ContainerNetwork;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.net.PemTrustOptions;
+import io.vertx.ext.consul.ConsulClient;
+import io.vertx.ext.consul.ConsulClientOptions;
+import io.vertx.ext.consul.dc.ConsulDatacenter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.consul.ConsulContainer;
+import org.testcontainers.utility.MountableFile;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Random;
+
+public class ConsulInstance extends ConsulContainer {
+  private static final Logger logger = LoggerFactory.getLogger(ConsulInstance.class);
+
+  private final ConsulDatacenter dc;
+
+  private JsonObject configuration = new JsonObject();
+
+  private ConsulInstance(String image, String version, ConsulDatacenter dc) {
+    super(image + ":" + version);
+    this.dc = dc;
+  }
+
+  public static ConsulInstance.Builder builder() {
+    return new Builder();
+  }
+
+  public static ConsulInstance.Builder defaultConsulBuilder(ConsulDatacenter dc) {
+    return ConsulInstance.builder()
+      .datacenter(dc)
+      .keyFile("server-key.pem")
+      .certFile("server-cert.pem")
+      .caFile("server-cert-ca-chain.pem");
+  }
+
+  @Override
+  public void stop() {
+    try {
+      execInContainer("consul", "leave");
+    } catch (IOException | InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+    super.stop();
+  }
+
+  public String getConfig(String name) {
+    return configuration.getString(name, "");
+  }
+
+  public void setConfig(JsonObject config) {
+    this.configuration = config;
+  }
+
+  public String address() {
+    if (isMacOS()) return getHost();
+    else return getContainerNetwork().getGateway();
+  }
+
+  public ConsulDatacenter dc() {
+    return dc;
+  }
+
+
+  public ConsulClient createClient(Vertx vertx, String token, boolean secure) {
+    return ConsulClient.create(vertx, consulClientOptions(token, secure));
+  }
+
+  public ConsulClient createClient(Vertx vertx, ConsulClientOptions options) {
+    return ConsulClient.create(vertx, options);
+  }
+
+  public ConsulClientOptions consulClientOptions(String token, boolean secure) {
+    return new ConsulClientOptions()
+      .setAclToken(token)
+      .setDc(dc.getName())
+      .setHost(address())
+      .setPort(secure ? getMappedPort(Builder.HTTPS_PORT) : getMappedPort(Builder.HTTP_PORT))
+      .setSsl(secure);
+  }
+
+  public ConsulClient createSecureClient(
+    Vertx vertx, String token, boolean trustAll, PemTrustOptions trustOptions
+  ) {
+    ConsulClientOptions options = consulClientOptions(token, true)
+      .setHost("127.0.0.1")
```

Comment:
why does this work with '127.0.0.1'?

Reply:
It's a default address. I fixed it, but there was an error "No subject alternative names matching IP address ...". Setting VerifyHost to false helped.